src/mem/secure.c

```diff
@@ -35,3 +45,35 @@ int mem_seccmp(const uint8_t *s1, const uint8_t *s2, size_t n)
 
 	return val;
 }
+
+
+#if !defined(__GNUC__) && !defined(USE_OPENSSL) && !defined(WIN32)
+/* Use a volatile pointer to memset to force the compiler always
+ * call it and not optimize away. */
+typedef void *(memset_t)(void *, int, size_t);
+static memset_t *const volatile memset_ptr = &memset;
+#endif
+
+/**
+ * Securely clean memory. This function is guaranteed not to get optimized
+ * away by compiler.
+ *
+ * @param data Pointer to data buffer
+ * @param size Size of the buffer
+ */
+void mem_secclean(void *data, size_t size)
+{
+#if defined(__GNUC__)
+	memset(data, 0, size);
+	/* Insert an asm statement that may potentially depend
+	* on the memory contents that were affected by memset.
+	* This prevents optimizing away the memset. */
+	__asm__ __volatile__("" : : "r" (data), "r" (size) : "memory");
+#elif defined(USE_OPENSSL)
+	OPENSSL_cleanse(data, size);
```

Comment:
Since OpenSSL 3 uses the same volatile memset pointer approach, this fallback has currently no benefits right? Should we replace it with memset_s instead?

Reply:
To my knowledge, `memset_s` is Microsoft's invention that practically only exists in MSVC. I don't see its benefits and I'd prefer not to use it.

As for OpenSSL, yes, recent versions use the volatile pointer approach, but older versions use different implementation. I think, they used asm and a loop of volatile stores at some point. If you want, I can remove it.

Reply:
Yes then lets remove OPENSSL_cleanse().

Reply:
Done.